### DISCUSSION TITLE
multiple robots discovery server config

### DISCUSSION THREAD

**Discussion by mdxtinkernick** [iRobotEducation/create3_docs#554] [Questions]
### How are you connecting to your Create 3?

USB Ethernet (ROS 2)

### Computer(s) Model(s) and Operating System(s)

pi4 ubuntu 2204 humble ros

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H2.2.3

### Which RMW is your robot running?

FastRTPS

### Does your robot have an assigned namespace? If so, please share.

no

### Is the robot connected to a network? If so, what is the network type?

isolated lan

### Are there multiple Create 3 robots connected to your network?

Yes

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

I am setting up multiple robots with discovery server running on pi connected to create3

I am not sure hoe to configure the discovery servers.

each pi/create3 will have a different ROS_DOMAIN_ID, and each pi will have a static ip on the wifi.

for the discovery servers (which will each be on a different ip address), should they also be on different ports for any reason (I don't think so, but am so confused now by all the permutations that I don't really know).

Does the remote server prefix for the discovery server have to be unique for each one, and if it does how do I configure the create3 against that - there is the option of using a custom xml file for the config on the base, but it would be useful to see what the default config is to see if it is using the number that all the examples seem to use, or the number is not used for that purpose

**Comment by slowrunner** [comment]
Hi @mdxtinkernick 

I am chasing the same questions about discovery server setup in a single Create3 robot setup.

I think there is guidance in this eProxima example on their discussion forum to the question "How to connect two discovery servers":

https://github.com/eProsima/Fast-DDS/discussions/3560#discussioncomment-6070579

I didn't examine it fully yet, but it does appear to mention configuring different Remote Server Prefixes, and more.

**Reply by mdxtinkernick** [reply]
Sorry, I wasn't clear in my description - I want each one to be isolated for now - will worry about them working together later. I think zenoh will be useful then, but I would like to get a class set working first as individual robots on a single network before I consider introducing another technology into the mix, each one with one desktop connected to it for RViz etc.

I think I have worked out the answer to my original question, am testing it this evening after tea. Its a shame things are so complicated and obscure. My understanding now is the third number in the remote server prefix is the guid and you specify it on the create base by adding ;'s before the details of the discovery server - number of ; is the id number - feels like a secret code that we are not supposed to be able to crack rather than an introduction to using ros2

**Reply by slowrunner** [reply]
> am testing it this evening after tea. Its a shame things are so complicated and obscure.

When you have this working, please post a drawing showing your environment, the servers (and invocation command), the robots, and please, please the various super_client files and shell environment variables for each student work station.  There are others in the same exact need, and the example will help those of us with slightly different needs to understand what we need to configure.

**Comment by slowrunner** [comment]
@mdxtinkernick 
> My understanding now is the third number in the remote server prefix is the guid and you specify it on the create base by adding ;'s before the details of the discovery server - number of ; is the id number - feels like a secret code that we are not supposed to be able to crack rather than an introduction to using ros2

The docs say the server id is the third number in the guid auto set - you do not need to add it on the command line:
Server 1:
```
<RemoteServer prefix="44.53.01.5f.45.50.52.4f.53.49.4d.41">
```

Server 0:
```
<RemoteServer prefix="44.53.00.5f.45.50.52.4f.53.49.4d.41">
```

Does that work to add semicolons in the Create3 enable FastDDS server editbox?  ```192.168.186.3:11811``` for server 0, ```;192.168.186.3:11811``` for server 1, ...?

Where the semicolons also come in is in the terminal export on the various desktops
ROS_DISCOVERY_SERVER="x.x.x.wifi_for_computeboard_of_create3_no_0:11811" to connect to discovery server 0  
ROS_DISCOVERY_SERVER=";x.x.x.wifi_for_computeboard_of_create3_no_1:11811" to connect to discovery server 1  
etc.

Where the semicolons come in also is in the terminal export on the various create3 compute boards 
ROS_DISCOVERY_SERVER="127.0.0.1:11811" to connect to discovery server 0  on create3 number “zero”
ROS_DISCOVERY_SERVER=“;127.0.0.1:11811" to connect to discovery server 1  on create3 number one
Etc
